feat: timezone conversion and assertion

atsyplenkov opened the pull request:
Hey @etiennebacher 👋

I decided to start by adding more support to `lubridate` package. I can see that it is a bit less structured than the other code in your package, so I am unsure how to best write it. If you are happy with the following approach, I will continue filling `tidypolars` with `lubridate` functions.

P.S. Note that I have added the `check_timezone` function, which throws an error earlier if the timezone code is not supported by `polars`. TIL that some `polars` functions allow a NULL timezone, while others do not.

etiennebacher commented:
Thanks, just so you know I won't be able to review this for a few days

etiennebacher commented on tests/testthat/test-funs_date.R:
```diff
@@ -560,3 +560,81 @@ test_that("ISOdatetime() works", {
   )
 })
+
+test_that("force_tz() works", {
+  test_df <- data.frame(
```
```suggestion
test_that("force_tz() works", {
  skip_if_not_installed("lubridate")
  test_df <- data.frame(
```

etiennebacher commented:
Hi @atsyplenkov, the PR looks good but I'd like the testing to be a bit more thorough. I have been adding more and more property-based testing to `tidypolars`, including for `lubridate` functions. Those can be found in `test-funs-date.R` while regular unit tests are in `test-funs_date.R` (I know, those names could be made more explicit).

If you're not familiar with property-based testing in R, I have a blog post on this: https://www.etiennebacher.com/posts/2024-10-01-using-property-testing-in-r/

For example, for `force_tz()`, here's what it looks like (combined with parametric testing using `patrick`):
```r
patrick::with_parameters_test_that(
  "extracting date components works", {
    for_all(
      tests = 20,
      datetime = posixct_(any_na = TRUE),
      property = function(datetime) {
        test_df <- data.frame(x1 = ymd_hms(datetime, tz = "UTC"))
        test <- pl$DataFrame(x1 = ymd_hms(datetime, tz = "UTC"))

        expect_equal_or_both_error(
          mutate(
            test,
            x1 = force_tz(x1, tz)
          ),
          mutate(
            test_df,
            x1 = force_tz(x1, tz)
          ),
          tolerance = 1e-6
        )
      }
    )
  },
  tz = list("Pacific/Auckland", "foobar", "", NA, c("Pacific/Auckland", "Africa/Abidjan"))
)
```
Running this may already uncover some bugs in the current implementation. It's more tedious to write and to pass but when it does then we can be confident that `tidypolars` and `lubridate` give the same results for a wide array of cases.

Could you rework the tests to include this type of testing? Don't hesitate to ask if you run into some issues while doing this. Note that if you need to convert some polars expressions to R values for the `tzone` argument then you can use `polars_expr_to_r()`.

etiennebacher commented on R/utils-expr.R:
```diff
@@ -963,3 +963,26 @@ polars_expr_to_r <- function(x) {
+#' @noRd
+#' @keywords internal
+check_timezone <- function(tz, null_allowed = FALSE) {
+  if (tz == "" & null_allowed) {
+    tz <- NULL
+    return(tz)
+  } else if (tz == "" & !null_allowed) {
```
Use `&&` instead of `&`. Also, I'm not sure how `polars` behave when there are several timezones e.g.

```r
dat <- data.frame( x = ymd_hms(rep("2009-08-07 00:00:01", 2), tz = "America/New_York"))
dat |>
  mutate(x2 = force_tz(x, c("UTC", "Pacific/Auckland")))
```

etiennebacher commented on R/utils-expr.R:
```diff
@@ -963,3 +963,26 @@ polars_expr_to_r <- function(x) {
+    tz <- NULL
+    return(tz)
+  } else if (tz == "" & !null_allowed) {
+    stop("This expession in `tidypolars` doesn't support NULL timezone")
+  } else if (tz != "") {
+    if (!any(tz == base::OlsonNames())) {
+      stop(sprintf("Unrecognized time zone '%s'", tz))
```
Use `abort()` instead of `stop()`

etiennebacher commented on R/utils-expr.R:
```diff
@@ -963,3 +963,26 @@ polars_expr_to_r <- function(x) {
+  } else if (tz != "") {
+    if (!any(tz == base::OlsonNames())) {
+      stop(sprintf("Unrecognized time zone '%s'", tz))
+    } else {
+      tz <- tz
+    }
```
This branch can be removed

atsyplenkov commented:
Thanks for the review. I am not familiar with property-based testing, but your blog post looks promising. I'll revise the PR during this week and come back to you

atsyplenkov commented:
> For example, for force_tz(), here's what it looks like (combined with parametric testing using patrick):

@etiennebacher I cannot implement property-based testing at this point. Even your example returns an error, which I cannot debug :( Do you have any idea why does it fail?



```r
patrick::with_parameters_test_that(
  "changing timezone works",
  {
    for_all(
      tests = 20,
      datetime = posixct_bounded(
        left = as.POSIXct("2000-01-01 00:00"),
        Sys.time(),
        any_na = FALSE
      ),
      property = function(datetime) {
        test_df <- data.frame(x1 = ymd_hms(datetime, tz = "UTC"))
        test <- as_polars_df(test_df)

        expect_equal(
          mutate(
            test,
            x1 = force_tz(time = x1, tzone = tz)
          ) |> as.data.frame(),
          mutate(
            test_df,
            x1 = force_tz(time = x1, tzone = tz)
          ) |> as.data.frame()
        )
      }
    )
  },
  tz = list("Pacific/Auckland")
)
```

```
Failure (test-funs-date.R:44:5): changing timezone works tz=Pacific/Auckland
Falsifiable after 1 tests, and 2 shrinks
<error/rlang_error>
Error in `mutate()`:
! Error while running function `force_tz()` in Polars.
x Invalid 'x' type in 'x && y'
---
Backtrace:
     x
  1. +-rlang::eval_tidy(code, args)
  2. \-quickcheck::for_all(...) at test-funs-date.R:44:5
  3.   \-hedgehog::forall(...)
  4.     \-hedgehog:::run.prop(property, counterexample$smallest, curry)
  5.       +-base::tryCatch(...)
  6.       | \-base (local) tryCatchList(expr, classes, parentenv, handlers)
  7.       |   \-base (local) tryCatchOne(expr, names, parentenv, handlers[[1L]])
  8.       |     \-base (local) doTryCatch(return(expr), name, parentenv, handler)
  9.       +-base::withCallingHandlers(...)
 10.       +-base::do.call(property, arguments)
 11.       \-tidypolars (local) `<fn>`(datetime = `<dttm>`)
 12.         +-tidypolars:::expect_equal(...) at test-funs-date.R:55:9
 13.         | \-base::inherits(x, "RPolarsDataFrame") at tests/testthat/helper.R:43:3
 14.         +-base::as.data.frame(mutate(test, x1 = force_tz(time = x1, tzone = tz)))
 15.         +-dplyr::mutate(test, x1 = force_tz(time = x1, tzone = tz))
 16.         \-tidypolars:::mutate.RPolarsDataFrame(test, x1 = force_tz(time = x1, tzone = tz))
Counterexample:
$datetime
[1] "2012-07-20 17:11:47 NZST"

Backtrace:
    ▆
 1. ├─rlang::eval_tidy(code, args)
 2. └─quickcheck::for_all(...) at test-funs-date.R:44:5
 3.   └─hedgehog::forall(...)
```

atsyplenkov commented:
@etiennebacher I cannot yet write the property-based testing yet, because the current implementation of `with_tz` and `force_tz` don't work when `tzone` argument passed externally, i.e.:

```r
library(tidyverse)
devtools::load_all()

TimeZone <- base::OlsonNames()[100]
TimeZone
#> "America/Creston"

test_df <- data.frame(
  dt_utc = c(
    ymd_hms(
      c(
        "2012-03-26 12:00:00",
        "2020-01-01 12:00:00",
        "2023-12-14 12:00:00",
        NA
      ),
      tz = "UTC"
    )
  )
)

as_polars_df(test_df) |>
  mutate(
    dt_utc = with_tz(dt_utc, "America/Creston")
  )

#> shape: (4, 1)
#> ┌───────────────────────────────┐
#> │ dt_utc                        │
#> │ ---                           │
#> │ datetime[ms, America/Creston] │
#> ╞═══════════════════════════════╡
#> │ 2012-03-26 05:00:00 MST       │
#> │ 2020-01-01 05:00:00 MST       │
#> │ 2023-12-14 05:00:00 MST       │
#> │ null                          │
#> └───────────────────────────────┘

as_polars_df(test_df) |>
  mutate(
    dt_utc = with_tz(dt_utc, TimeZone)
  )

#> Error in `mutate()`:
#>   ! Error while running function `with_tz()` in Polars.
#>   ✖ Invalid 'x' type in 'x && y'
#>   Hide Traceback
#>        ▆
#>     1. ├─dplyr::mutate(...)
#>     2. └─tidypolars:::mutate.RPolarsDataFrame(as_polars_df(test_df), dt_utc = with_tz(dt_utc, TimeZone))
#>     3.   └─tidypolars:::translate_dots(...) at tidypolars/R/mutate.R:117:3
#>     4.     └─base::lapply(...) at tidypolars/R/utils-expr.R:36:3
#>     5.       └─tidypolars (local) FUN(X[[i]], ...)
#>     6.         └─tidypolars:::translate_expr(...) at tidypolars/R/utils-expr.R:46:5
#>     7.           └─tidypolars:::translate(...) at tidypolars/R/utils-expr.R:145:5
#>     8.             └─base::tryCatch(...) at tidypolars/R/utils-expr.R:622:7
#>     9.               └─base (local) tryCatchList(expr, classes, parentenv, handlers)
#>    10.                 └─base (local) tryCatchOne(expr, names, parentenv, handlers[[1L]])
#>    11.                   └─value[[3L]](cond)
#>    12.                     └─rlang::abort(...) at tidypolars/R/utils-expr.R:638:13

```

I cannot wrap my head around on how to solve this issue.

etiennebacher commented:
The issue is that when we handle `tz` in `check_timezone()`, it has already been converted to a polars expression, e.g. `pl$lit("Europe/London")`. Therefore we cannot use `length(tz)` for instance. The solution is to convert this back to an R expression using `polars_expr_to_r()`. This is a bit unfortunate but I don't have  better solution for now because some function arguments need to be translated (e.g. `x` in `mean(x, na.rm = TRUE)`) but others don't (e.g. `na.rm` in `mean(x, na.rm = TRUE)`), and I don't have a way of knowing this in advance.

I also just realized that `lubridate` allows several timezones in a single column, for instance by passing `tz = c("Europe/London", "Pacific/Auckland")`, but Polars doesn't. I have added a message for this, and it should be mentioned in the [table notes](https://www.tidypolars.etiennebacher.com/articles/r-and-polars-expressions#list-of-base-r-and-tidyverse-functions-supported-by-tidypolars).

This now works:
```r
patrick::with_parameters_test_that(
  "changing timezone works", {
    for_all(
      tests = 20,
      datetime = posixct_(any_na = TRUE),
      property = function(datetime) {
        test_df <- data.frame(x1 = ymd_hms(datetime, tz = "UTC"))
        test <- pl$DataFrame(x1 = ymd_hms(datetime, tz = "UTC"))

        expect_equal_or_both_error(
          mutate(
            test,
            x1 = force_tz(x1, tz)
          ),
          mutate(
            test_df,
            x1 = force_tz(x1, tz)
          ),
          tolerance = 1e-6
        )
      }
    )
  },
  tz = list("Pacific/Auckland", "foobar", "", NA)
)
```

Can you also add tests for the various error messages in `check_timezone()` and for `with_tz()`?

atsyplenkov commented:
Thanks! I'll do it.

etiennebacher commented:
@atsyplenkov just FYI I plan on doing a new release before mid-March so let me know if you need some help finishing this PR before that.

etiennebacher commented:
No worries, I'm not in a rush at all, I just wanted to inform you in case those functions are important in some of your project . Take your time, good luck with work :slightly_smiling_face:

atsyplenkov commented:
Hey, just a quick update—I haven't forgotten about this PR, I'm just still extremely busy.

etiennebacher commented:
No worries, finish when you can. I'll make a new release since there are interesting fixes and features I think.

atsyplenkov commented on tests/testthat/test-funs-date.R:
```diff
@@ -37,3 +40,62 @@ patrick::with_parameters_test_that(
+    )
+  },
+  # TODO:
+  # Add support for unrecognized timezone ("foo") and NULL ("")
```
Cannot wrap my head around how to properly handle it. From my perspective, the safest approach  — just skip

etiennebacher replied:
Thanks for opening the upstream issue in `lubridate`, I think this should error so polars does the right thing IMO. Let's uncomment this if `lubridate` changes the behavior to error instead.

atsyplenkov commented:
Gee, what a rabbit hole of the unrecognized timezone handling. TBH, I don't know what to do.

The problem is that `polars` does not allow replacing the timezone with a gibberish, while `lubridate` (and R in general) allows. At the same time, `with_tz()` can handle it, but `force_tz()` cannot handle (see issue https://github.com/tidyverse/lubridate/issues/1186).

Therefore, I am keen to throw an early error when an unrecognized timezone is passed. This will be reflected in the number of property-based tests we can implement. See my comment there

Apart from that (sorry for the useless commit name, I accidentally hit enter too early):
- [x] added tests for the various error messages in `check_timezone()`
- [x] added property tests for `with_tz()` and `force_tz()`
- [x] updated the documentation.

atsyplenkov commented on tests/testthat/test-funs-date.R:
```diff
@@ -1,3 +1,6 @@
+# TODO:
+# Sometimes this test fails, e.g. try seed == 1
+set.seed(123)
```
On my machine, the "extracting date components works" test is failing with `set.seed(1)`. Can you check if it the case with you as well?

etiennebacher replied:
I think this is an error in polars:

``` r
date <- as.POSIXct("2493-05-15 00:00:00 CEST")

data.frame(x1 = date)
#>           x1
#> 1 2493-05-15
polars::pl$DataFrame(x1 = date)
#> shape: (1, 1)
#> ┌─────────────────────┐
#> │ x1                  │
#> │ ---                 │
#> │ datetime[ms]        │
#> ╞═════════════════════╡
#> │ 2493-05-14 23:00:00 │ # <<<<<<<<< shows 2493-05-14 and not 2493-05-15
#> └─────────────────────┘
```

atsyplenkov replied:
Good to know, you've got an eagle eye!

etiennebacher commented on tests/testthat/test-funs-date-lazy.R:
```diff
@@ -102,7 +102,7 @@ patrick::with_parameters_test_that(
   # TODO: see https://github.com/tidyverse/lubridate/issues/1186
   # If lubridate changed the behavior to error, then we can add "foo" and ""
-  tz = list("Pacific/Auckland", "US/Alaska", NA)
+  tz = list("Pacific/Auckland", "Europe/Rome", NA)
```
Looks like "US/Alaska" in not in `OlsonNames()`

atsyplenkov commented:
Awesome, thanks for your guidance and help! 

Just to let you know, I started wrapping some other `lubridate` functions. Hope that their implementation will go smoother

etiennebacher commented:
Great, looking forward to seeing them!
